### PR TITLE
[4669] Allow users to change routes and courses on registered trainees

### PR DESCRIPTION
**Pull request by HettieS**
### Context

This branch includes the new training routes stuff, plus fixes for the below bugs:
https://trello.com/c/fHrmVh7z/5019-funding-method-not-being-cleared-for-publish-course-changes
https://trello.com/c/GbgVm6Lb/5050-production-changing-from-one-publish-course-to-another-publish-course-does-not-update-the-trainees-course-details

Trello: https://trello.com/c/94rk0ez9/4669-l-allow-users-to-change-routes-and-courses-on-registered-trainees

We are allowing registered trainees to amend their training route via the Change course flow. On the course details (for both draft and registered trainees) the route is now displayed as part of the course section, and if you click change, you will be taken to the training routes form and then onto the relevant course flow.

/trainees/{trainee_id}/course-details/confirm
Screenshot 2022-10-12 at 10 22 27

If you click to change course, you will first be displayed with the training route page which now carries a context:
/trainees/{trainee_id}/training-routes/edit?context=edit-course

Screenshot 2022-10-12 at 10 30 47

The original functionality for amending the training route on draft trainees has not changed:
/trainees/{trainee_id}/review-draft
Screenshot 2022-10-12 at 10 29 25

When the new training route is confirmed we use the existing RouteDataManager to determine what needs to be deleted.

As part of this PR I have not considered removing data that we no longer need e.g. lead schools when the route changed to from school direct to a different route. I would suggest we do this as a follow up piece for v2.

### Changes proposed in this pull request

* Add new TrainingRouteForm and update TrainingRouteController
* Add def training_route to BaseController
* Update RouteDataManager so that it does not send Trainee::Update (this happens in the controller)
* Update PublishCourseDetailsController and PublishCourseDetailsForm
* Update def available_courses on the Trainee model as we no longer show all courses for Apply apps
* Update TrainingRouteManager to expose training_route
* Update TraineeHelper
* Update CourseDetailsForm
* Update ConfirmCourseDetails, which is for Apply trainees
* Update IttDatesForm with TrainingRoutesForm instance
* Add page objects
* Bug fix where we were not changing publish course details properly/amending the allocation subject/clearing funding properly

### Guidance to review

* For different training routes, check that you are redirected on to the correct course options
* Check the route and course change behaviour for both draft and registered trainees (we only stash on registered trainees and not on drafts)
* Make sure we're clearing funding info in the way that is expected
* Check backlinks because i think some of those might not be quite right
* Check the existing functionality for drafts still works
* Check the content makes sense for the new flow (mainly around Apply applications) as I might have missed this

For review bug fixes:
* Switch from a non-funding to a funded publish course for a registered trainee - observe it updates the course correctly and the funding is cleared
* Try the other way around too (funded to non-funded)

### Important business

- [ ] ~~Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?~~
- [ ] ~~Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?~~

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: app/forms/publish_course_details_form.rb [hunk 1]
```diff
@@ -53,6 +58,7 @@ def save!
     return false unless valid?
 
     update_trainee_attributes
+    clear_funding_information if clear_funding_information?
```
[the remark this answers is not in the document]
Comment (HettieS) [reply app/forms/publish_course_details_form.rb:61]: I've added this check here as it is used in CourseDetailsForm and ConfirmCourseForm and seems it was missing here - I don't think we were clearing funding properly for publish courses potentially

**Review by jhenderson — CHANGES_REQUESTED** [review]
Great work @HettieS. A few small suggestions.

File: app/forms/publish_course_details_form.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (HettieS) [reply app/forms/publish_course_details_form.rb]: I've removed this and added it to another branch here we should do the fix for this.

**Comment by HettieS** [comment]
> I've just seen a bug where we clear funding when the route/subject changes but we then don't surface the correct funding option - we give them a page relating to the previous course. Not sure if existing bug or created in this PR yet. Screenshot 2022-10-12 at 11 41 20 Screenshot 2022-10-12 at 11 41 28

note: I've raised this as a separate bug ticket and will not fix on this PR

File: app/helpers/publish_courses_helper.rb [hunk 2]
```diff
@@ -9,14 +9,7 @@ def apply_course_and_route_summary(course)
...
-  def course_summary_text_for(trainee, course)
-    summary_with_route = t(".summary_with_route", summary: course.summary, route: t("activerecord.attributes.trainee.training_routes.#{course.route}"))
-    trainee.apply_application? ? summary_with_route : course.summary
-  end
-
-  def courses_fieldset_text_for(trainee)
-    return t("views.forms.publish_course_details.all_courses_message") if trainee.apply_application?
-
-    t("views.forms.publish_course_details.route_message", route: route_title(trainee.training_route))
+  def courses_fieldset_text_for(_trainee)
```
Comment (jhenderson) [inline app/helpers/publish_courses_helper.rb:12]: You can get rid of the `_trainee` argument. You'll have to update the method here also: http://github.com/DFE-Digital/register-trainee-teachers/blob/f2f6f3a83e5be4b366629cd2ca09a868d2bec962/app/views/trainees/publish_course_details/edit.html.erb#L29-L29

File: app/views/trainees/publish_course_details/edit.html.erb [hunk 3]
```diff
@@ -31,7 +31,7 @@
           <% @courses.each_with_index do |course, i| %>
             <%= f.govuk_radio_button :course_uuid, course.uuid,
               label: { text: "#{course.name} (#{course.code})" },
-              hint: { text: course_summary_text_for(@trainee, course) }, 
+              hint: { text: "#{course.summary}" },
```
Comment (jhenderson) [inline app/views/trainees/publish_course_details/edit.html.erb:34]:
```suggestion
              hint: { text: course.summary },
```

**Review by jhenderson — APPROVED** [review]
Looking good. Some minor suggestions.

**Comment by HettieS** [comment]
Existing bugs in the publish course details flow/clearing the funding should now be fixed on this PR.

**Comment by HettieS** [comment]
[Production: changing from one publish course to another publish course does not update the trainee's course details](https://trello.com/c/GbgVm6Lb/5050-production-changing-from-one-publish-course-to-another-publish-course-does-not-update-the-trainees-course-details)

**Comment by HettieS** [comment]
[Funding method not being cleared for publish course changes](https://trello.com/c/fHrmVh7z/5019-funding-method-not-being-cleared-for-publish-course-changes)

**Review by edwardhorsford — APPROVED** [review]
Have walked through this in detail with @HettieS. Looking great!

We've found a number of bugs along the way - but believe they are all preexisting issues so we'll raise separately.

🎉

**Review by markhallen — APPROVED** [review]
LGTM

**Pull request merged**